<!DOCTYPE html>
<html lang="en-us">
  <head>
    
<meta charset="UTF-8">
<title>Unexpected exceptions | Elasticsearch .NET Clients [7.17] | Elastic</title>
<meta class="elastic" name="content" content="Unexpected exceptions | Elasticsearch .NET Clients [7.17]">

<link rel="home" href="index.html" title="Elasticsearch .NET Clients [7.17]"/>
<link rel="up" href="thrown-exceptions.html" title="Exceptions"/>
<link rel="prev" href="thrown-exceptions.html" title="Exceptions"/>
<link rel="next" href="unrecoverable-exceptions.html" title="Unrecoverable exceptions"/>
<meta class="elastic" name="product_version" content="7.17"/>
<meta class="elastic" name="product_name" content="Clients"/>
<meta class="elastic" name="website_area" content="documentation"/>
<meta name="DC.type" content="Learn/Docs/Clients/.Net/7.17"/>
<meta name="DC.subject" content="Clients"/>
<meta name="DC.identifier" content="7.17"/>
<meta name="robots" content="noindex,nofollow"/>

    <meta http-equiv="content-type" content="text/html; charset=utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <script src="https://cdn.optimizely.com/js/18132920325.js"></script>
    <link rel="icon" href="/favicon.ico" sizes="32x32"/>
    <link rel="icon" href="/favicon.svg" type="image/svg+xml"/>
    <link rel="apple-touch-icon" href="/apple-touch-icon.png" sizes="180x180"/>
    <link rel="manifest" href="/manifest.json">
    <meta name="apple-mobile-web-app-title" content="Elastic">
    <meta name="application-name" content="Elastic">
    <meta name="msapplication-TileColor" content="#ffffff">
    <meta name="msapplication-TileImage" content="/mstile-144x144.png">
    <meta name="theme-color" content="#ffffff">
    <meta name="naver-site-verification" content="936882c1853b701b3cef3721758d80535413dbfd" />
    <meta name="yandex-verification" content="d8a47e95d0972434" />
    <meta name="localized" content="true" />
    <meta name="st:robots" content="follow,index" />
    <meta property="og:image" content="https://static-www.elastic.co/v3/assets/bltefdd0b53724fa2ce/blt280217a63b82a734/6202d3378b1f312528798412/elastic-logo.svg" />
    <meta property="og:image:width" content="500" />
    <meta property="og:image:height" content="172" />
    <!-- Give IE8 a fighting chance -->
    <!--[if lt IE 9]>
    <script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
    <script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
    <![endif]-->
    <link rel="stylesheet" type="text/css" href="/guide/static/styles-v1.css" />
  </head>

  <!--© 2015-2025 Elasticsearch B.V. -->
  <!-- All Elastic documentation is licensed under a Creative Commons Attribution-NonCommercial-NoDerivatives 4.0 International License. -->
  <!-- http://creativecommons.org/licenses/by-nc-nd/4.0/ -->

  <body>
    <!-- Google Tag Manager -->
    <script>dataLayer = [];</script><noscript><iframe src="//www.googletagmanager.com/ns.html?id=GTM-58RLH5" height="0" width="0" style="display:none;visibility:hidden"></iframe></noscript>
    <script>(function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start': new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0], j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src= '//www.googletagmanager.com/gtm.js?id='+i+dl;f.parentNode.insertBefore(j,f); })(window,document,'script','dataLayer','GTM-58RLH5');</script>
    <!-- End Google Tag Manager -->

    <!-- Global site tag (gtag.js) - Google Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-12395217-16"></script>
    <script>
      window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments);}
      gtag('js', new Date());
      gtag('config', 'UA-12395217-16');
    </script>

    <!-- Google Tag Manager for GA4 -->
    <script>(function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start': new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0], j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src='https://www.googletagmanager.com/gtm.js?id='+i+dl;f.parentNode.insertBefore(j,f);})(window,document,'script','dataLayer','GTM-KNJMG2M');</script>
    <noscript><iframe src="https://www.googletagmanager.com/ns.html?id=GTM-KNJMG2M" height="0" width="0" style="display:none;visibility:hidden"></iframe></noscript>
    <!-- End Google Tag Manager for GA4-->

    <div id='elastic-nav' style="display:none;"></div>
    <script src='https://www.elastic.co/elastic-nav.js'></script>

    <div class="main-container">
      <section id="content" >
        <div class="content-wrapper">

          <section id="guide" lang="en">
            <div class="container-fluid">
              <div class="row pb-3">
                <div class="col-12 order-2 col-md-4 order-md-1 col-lg-3 h-almost-full-md sticky-top-md" id="left_col">
                  <!-- The TOC is appended here -->
                </div>

                <div class="col-12 order-1 col-md-8 order-md-2 col-lg-7 order-lg-2 guide-section" id="middle_col">
                  <!-- start body -->
                  
<div class="page_header">
<strong>IMPORTANT</strong>: No additional bug fixes or documentation updates
will be released for this version. For the latest information, see the
<a href="../current/index.html">current release documentation</a>.
</div>
<div class="navheader">
<span class="prev">
<a href="thrown-exceptions.html">« Exceptions</a>
</span>
<span class="next">
<a href="unrecoverable-exceptions.html">Unrecoverable exceptions »</a>
</span>
</div>
<div class="book" lang="en">
<div class="titlepage">
<div class="breadcrumbs">
<span class="breadcrumb-link"><a href="/guide/">Elastic Docs</a></span>
<span class="chevron-right">›</span><span class="breadcrumb-link"><a href="index.html">Elasticsearch .NET Clients [7.17]</a></span>
<span class="chevron-right">›</span><span class="breadcrumb-link"><a href="api-conventions.html">API Conventions</a></span>
<span class="chevron-right">›</span><span class="breadcrumb-link"><a href="thrown-exceptions.html">Exceptions</a></span>
</div>
<div>
<div><h1 class="title"><a id="id-1"></a>Unexpected exceptions</h1><a class="edit_me edit_me_private" rel="nofollow" title="Editing on GitHub is available to Elastic" href="https://github.com/elastic/elasticsearch-net/edit/7.17/docs/client-concepts/connection-pooling/exceptions/unexpected-exceptions.asciidoc">edit</a></div>
</div>
<!--EXTRA-->
</div>
<div id="content">
<div id="url-to-v3" class="version-warning">
    <strong>IMPORTANT</strong>: This documentation is no longer updated. Refer to <a href="https://www.elastic.co/support/eol">Elastic's version policy</a> and the <a href="https://www.elastic.co/docs">latest documentation</a>.
</div>
<div class="section">
<div class="titlepage"><div><div>
<div class="position-relative"><h2 class="title"><a id="unexpected-exceptions"></a>Unexpected exceptions</h2><a class="edit_me edit_me_private" rel="nofollow" title="Editing on GitHub is available to Elastic" href="https://github.com/elastic/elasticsearch-net/edit/7.17/docs/client-concepts/connection-pooling/exceptions/unexpected-exceptions.asciidoc">edit</a></div>
</div></div></div>
<p>When a client call throws an exception that the <code class="literal">IConnection</code> cannot handle, the exception will bubble
out of the client as an <code class="literal">UnexpectedElasticsearchClientException</code>, regardless of whether the client is configured to
throw exceptions or not.</p>
<p>An <code class="literal">IConnection</code> is in charge of knowing which exceptions it can recover from and those it can&#8217;t, and the default <code class="literal">IConnection</code>
in</p>
<div class="ulist itemizedlist">
<ul class="itemizedlist">
<li class="listitem">
Desktop CLR is based on <code class="literal">WebRequest</code> which can and will recover from <code class="literal">WebException</code>
</li>
<li class="listitem">
Core CLR is based on <code class="literal">HttpClient</code> which can and will recover from <code class="literal">HttpRequestException</code>
</li>
</ul>
</div>
<p>Other exceptions will be grounds for immediately exiting the pipeline. Let&#8217;s demonstrate this
using our Virtual cluster test framework</p>
<div class="pre_wrapper lang-csharp">
<div class="console_code_copy" title="Copy to clipboard"></div>
<pre class="programlisting prettyprint lang-csharp">var audit = new Auditor(() =&gt; VirtualClusterWith <a id="CO2-1"></a><i class="conum" data-value="1"></i>
    .Nodes(10)
    .ClientCalls(r =&gt; r.SucceedAlways())
    .ClientCalls(r =&gt; r.OnPort(9201).FailAlways(new Exception("boom!"))) <a id="CO2-2"></a><i class="conum" data-value="2"></i>
    .StaticConnectionPool()
    .Settings(s =&gt; s.DisablePing())
);

audit = await audit.TraceCall(
    new ClientCall {
        { AuditEvent.HealthyResponse, 9200 }, <a id="CO2-3"></a><i class="conum" data-value="3"></i>
    }
);

audit = await audit.TraceUnexpectedException(
    new ClientCall {
        { AuditEvent.BadResponse, 9201 }, <a id="CO2-4"></a><i class="conum" data-value="4"></i>
    },
    (e) =&gt;
    {
        e.FailureReason.Should().Be(PipelineFailure.Unexpected);
        e.InnerException.Should().NotBeNull();
        e.InnerException.Message.Should().Be("boom!");
    }
);</pre>
</div>
<div class="calloutlist">
<table border="0" summary="Callout list">
<tr>
<td align="left" valign="top" width="5%">
<p><a href="#CO2-1"><i class="conum" data-value="1"></i></a></p>
</td>
<td align="left" valign="top">
<p>set up a cluster with 10 nodes</p>
</td>
</tr>
<tr>
<td align="left" valign="top" width="5%">
<p><a href="#CO2-2"><i class="conum" data-value="2"></i></a></p>
</td>
<td align="left" valign="top">
<p>where node 2 on port 9201 always throws an exception</p>
</td>
</tr>
<tr>
<td align="left" valign="top" width="5%">
<p><a href="#CO2-3"><i class="conum" data-value="3"></i></a></p>
</td>
<td align="left" valign="top">
<p>The first call to 9200 returns a healthy response</p>
</td>
</tr>
<tr>
<td align="left" valign="top" width="5%">
<p><a href="#CO2-4"><i class="conum" data-value="4"></i></a></p>
</td>
<td align="left" valign="top">
<p>&#8230;&#8203;but the second call, to 9201, returns a bad response</p>
</td>
</tr>
</table>
</div>
<p>Sometimes, an unexpected exception happens further down in the pipeline. In this scenario, we
wrap them inside an <code class="literal">UnexpectedElasticsearchClientException</code> so that information about where
in the pipeline the exception happened is not lost.</p>
<p>In this next example, a call to 9200 fails with a <code class="literal">WebException</code>.
The call then rolls over to retry on 9201, which throws an hard exception from within <code class="literal">IConnection</code>.
Finally, we assert that we can still see the audit trail for the whole coordinated request.</p>
<div class="pre_wrapper lang-csharp">
<div class="console_code_copy" title="Copy to clipboard"></div>
<pre class="programlisting prettyprint lang-csharp">var audit = new Auditor(() =&gt; VirtualClusterWith
    .Nodes(10)
#if DOTNETCORE
    .ClientCalls(r =&gt; r.OnPort(9200).FailAlways(new System.Net.Http.HttpRequestException("recover"))) <a id="CO3-1"></a><i class="conum" data-value="1"></i>
#else
    .ClientCalls(r =&gt; r.OnPort(9200).FailAlways(new System.Net.WebException("recover"))) <a id="CO3-2"></a><i class="conum" data-value="1"></i>
#endif
    .ClientCalls(r =&gt; r.OnPort(9201).FailAlways(new Exception("boom!"))) <a id="CO3-3"></a><i class="conum" data-value="2"></i>
    .StaticConnectionPool()
    .Settings(s =&gt; s.DisablePing())
);

audit = await audit.TraceUnexpectedException(
    new ClientCall {
        { AuditEvent.BadResponse, 9200 },
        { AuditEvent.BadResponse, 9201 }, <a id="CO3-4"></a><i class="conum" data-value="3"></i>
    },
    (e) =&gt;
    {
        e.FailureReason.Should().Be(PipelineFailure.Unexpected);
        e.InnerException.Should().NotBeNull();
        e.InnerException.Message.Should().Be("boom!");
    }
);</pre>
</div>
<div class="calloutlist">
<table border="0" summary="Callout list">
<tr>
<td align="left" valign="top" width="5%">
<p><a href="#CO3-1"><i class="conum" data-value="1"></i></a><a href="#CO3-2"></a></p>
</td>
<td align="left" valign="top">
<p>calls on 9200 set up to throw a <code class="literal">HttpRequestException</code> or a <code class="literal">WebException</code></p>
</td>
</tr>
<tr>
<td align="left" valign="top" width="5%">
<p><a href="#CO3-3"><i class="conum" data-value="2"></i></a></p>
</td>
<td align="left" valign="top">
<p>calls on 9201 set up to throw an <code class="literal">Exception</code></p>
</td>
</tr>
<tr>
<td align="left" valign="top" width="5%">
<p><a href="#CO3-4"><i class="conum" data-value="3"></i></a></p>
</td>
<td align="left" valign="top">
<p>Assert that the audit trail for the client call includes the bad response from 9200 and 9201</p>
</td>
</tr>
</table>
</div>
<p>An unexpected hard exception on ping and sniff is something we <span class="strong strong"><strong>do</strong></span> try to recover from and failover to retrying on the next node.</p>
<p>Here, pinging nodes on first use is enabled and the node on port 9200 throws an exception on ping; when this happens,
we still fallover to retry the ping on node on port 9201, where it succeeds.
Following this, the client call on 9201 throws a hard exception that we are not able to recover from.</p>
<div class="pre_wrapper lang-csharp">
<div class="console_code_copy" title="Copy to clipboard"></div>
<pre class="programlisting prettyprint lang-csharp">var audit = new Auditor(() =&gt; VirtualClusterWith
    .Nodes(10)
    .Ping(r =&gt; r.OnPort(9200).FailAlways(new Exception("ping exception")))
    .Ping(r =&gt; r.OnPort(9201).SucceedAlways())
    .ClientCalls(r =&gt; r.OnPort(9201).FailAlways(new Exception("boom!")))
    .StaticConnectionPool()
    .AllDefaults()
);

audit = await audit.TraceUnexpectedException(
    new ClientCall {
        { AuditEvent.PingFailure, 9200 },
        { AuditEvent.PingSuccess, 9201 },
        { AuditEvent.BadResponse, 9201 },
    },
    e =&gt;
    {
        e.FailureReason.Should().Be(PipelineFailure.Unexpected);

        e.InnerException.Should().NotBeNull();
        e.InnerException.Message.Should().Be("boom!"); <a id="CO4-1"></a><i class="conum" data-value="1"></i>

        e.SeenExceptions.Should().NotBeEmpty(); <a id="CO4-2"></a><i class="conum" data-value="2"></i>
        var pipelineException = e.SeenExceptions.First();
        pipelineException.FailureReason.Should().Be(PipelineFailure.PingFailure);
        pipelineException.InnerException.Message.Should().Be("ping exception");

        var pingException = e.AuditTrail.First(a =&gt; a.Event == AuditEvent.PingFailure).Exception; <a id="CO4-3"></a><i class="conum" data-value="3"></i>
        pingException.Should().NotBeNull();
        pingException.Message.Should().Be("ping exception");
    }
);</pre>
</div>
<div class="calloutlist">
<table border="0" summary="Callout list">
<tr>
<td align="left" valign="top" width="5%">
<p><a href="#CO4-1"><i class="conum" data-value="1"></i></a></p>
</td>
<td align="left" valign="top">
<p><code class="literal">InnerException</code> is the exception that brought the request down</p>
</td>
</tr>
<tr>
<td align="left" valign="top" width="5%">
<p><a href="#CO4-2"><i class="conum" data-value="2"></i></a></p>
</td>
<td align="left" valign="top">
<p>The hard exception that happened on ping is still available though</p>
</td>
</tr>
<tr>
<td align="left" valign="top" width="5%">
<p><a href="#CO4-3"><i class="conum" data-value="3"></i></a></p>
</td>
<td align="left" valign="top">
<p>An exception can be hard to relate back to a point in time, so the exception is also available on the audit trail</p>
</td>
</tr>
</table>
</div>
</div>
</div>
</div><div class="navfooter">
<span class="prev">
<a href="thrown-exceptions.html">« Exceptions</a>
</span>
<span class="next">
<a href="unrecoverable-exceptions.html">Unrecoverable exceptions »</a>
</span>
</div>

                  <!-- end body -->
                </div>

                <div class="col-12 order-3 col-lg-2 order-lg-3 h-almost-full-lg sticky-top-lg" id="right_col">
                  <div id="sticky_content">
                    <!-- The OTP is appended here -->
                    <div class="row">
                      <div class="col-0 col-md-4 col-lg-0" id="bottom_left_col"></div>
                      <div class="col-12 col-md-8 col-lg-12">
                        <div id="rtpcontainer">
                          <div class="mktg-promo" id="most-popular">
                            <p class="aside-heading">Most Popular</p>
                            <div class="pb-2">
                              <p class="media-type">Video</p>
                              <a href="https://www.elastic.co/webinars/getting-started-elasticsearch?page=docs&placement=top-video">
                                <p class="mb-0">Get Started with Elasticsearch</p>
                              </a>
                            </div>
                            <div class="pb-2">
                              <p class="media-type">Video</p>
                              <a href="https://www.elastic.co/webinars/getting-started-kibana?page=docs&placement=top-video">
                                <p class="mb-0">Intro to Kibana</p>
                              </a>
                            </div>
                            <div class="pb-2">
                              <p class="media-type">Video</p>
                              <a href="https://www.elastic.co/webinars/introduction-elk-stack?page=docs&placement=top-video">
                                <p class="mb-0">ELK for Logs & Metrics</p>
                              </a>
                            </div>
                          </div>
                        </div>

                        <!-- Feedback widget -->
                        <div id="feedbackWidgetContainer"></div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </section>

        </div>


        <div id='elastic-footer'></div>
        <script src='https://www.elastic.co/elastic-footer.js'></script>
        <!-- Footer Section end-->

      </section>
    </div>

    <!-- Feedback modal -->
    <div id="feedbackModalContainer"></div>

    <script src="/guide/static/jquery.js"></script>
    <script type="text/javascript" src="/guide/static/docs-v1.js"></script>
    <script type="text/javascript">
  window.initial_state = {}</script>
  </body>
</html>
